### ISSUE TITLE
group_keypoints

### ISSUE THREAD

**Issue by XiaoJiNu** [Daniil-Osokin/lightweight-human-pose-estimation.pytorch#8]
could you please explain the process of group_keypoints?

**Comment by Daniil-Osokin** [comment]
Network has two output:
1. Heatmaps to extract keypoint coordinates at local maximums.
2. "Part affinity fields", which help in groupping.

There is pre-defined ```set``` of keypoint pairs: left shoulder and left elbow, left elbow and left wrist, neck and left shoulder, and so on. Each 2 channels (one for *dx*, one for *dy*) in the second tensor learn oriented connection (normalized vector) between keypoints of specific person, for particular keypoint pair from the ```set```. So ground truth for learning is generated in the following way: for each person take keypoint pair from the ```set```, compute vector between keypoints (as coordinates difference), normalize it and place this values (*dx*, *dy*) in the corresponding cells of channels in ground truth tensor along the line between keypoints. During inference exhaustive serach is performed:
* For each keypoint pair from the ```set```:
  * Match each found keypoint of type of the first keypoint from the pair (e.g left shoulder) with each found keypoint of type of the second keypoint from the pair (e.g. left elbow). Keypoints matched (belong to the same person), if vector between them corresponds to vector, learned by the second ("part affinity fields") tensor. Correspondence is verified (by cosine distance) along the line between keypoints.

**Comment by XiaoJiNu** [comment]
Thanks for your reply,  i have read the code and now i'm confusing about the ralationship between BODY_PARTS_KPT_IDS and BODY_PARTS_PAF_IDS.  
BODY_PARTS_KPT_IDS = [[1, 2], [1, 5], [2, 3], [3, 4], [5, 6], [6, 7], [1, 8], [8, 9], [9, 10], [1, 11],
                      [11, 12], [12, 13], [1, 0], [0, 14], [14, 16], [0, 15], [15, 17], [2, 16], [5, 17]]
BODY_PARTS_PAF_IDS = ([12, 13], [20, 21], [14, 15], [16, 17], [22, 23], [24, 25], [0, 1], [2, 3], [4, 5],
                      [6, 7], [8, 9], [10, 11], [28, 29], [30, 31], [34, 35], [32, 33], [36, 37], [18, 19], [26, 27])

**Comment by XiaoJiNu** [comment]
hi Daniil-Osokin @Daniil-Osokin , i have another question.
At the beginning,  ratio = passed_point_score / passed_point_num is the average score of passed point.But what does ratio += min(height_n / vec_norm - 1, 0)  mean?

**Comment by Daniil-Osokin** [comment]
So, `BODY_PARTS_KPT_IDS` contains keypoint id pairs, which form pre-defined `set` of keypoint pairs, between them part affinity fields are learned.  `BODY_PARTS_PAF_IDS` contains pairs of channels id (for *dx*, *dy*) in part affinity fields tensor for corresponding keypoint pair.

The `min(height_n / vec_norm - 1, 0)` is the penalty on the large distance between keypoints of the same person. If, e.g. distance between left shoulder and left elbow higher than half of the frame, then we penalize connection between this keypoints (decrease the average score, and if score is weak, it will not pass the threshold).

**Comment by XiaoJiNu** [comment]
@Daniil-Osokin thank you very much, i understand

**Comment by Daniil-Osokin** [comment]
You are welcome!

**Comment by anhphi257** [comment]
Hi @Daniil-Osokin , i want to understand deeper about how to design the BODY_PARTS_PAF_IDS: 
1. How to design the index?
2. How to design the relation between them and use them for training?

**Comment by Daniil-Osokin** [comment]
There is no magic here. First, you define order of `paf` output (which channels from output tensor corresponds to `paf` between specific keypoints) when preparing ground truth for them. Then in inference define keypoint pairs (`BODY_PARTS_KPT_IDS`) and make pairs of channels (`BODY_PARTS_PAF_IDS`) from `paf` output in order corresponding to `BODY_PARTS_KPT_IDS`, so take the item from `BODY_PARTS_KPT_IDS`, find which channels in `paf` output corresponds to connection between these keypoints and add such channel indices into `BODY_PARTS_PAF_IDS`.
